File: internal/client/message.go
package client

import (
	"strings"
)

type RawMessage struct {
	Prefix  string
	Command string
	Params  []string
	Text    string
}

func parseRawMessage(s string) RawMessage {
	var (
		message = strings.Split(strings.TrimSpace(s), " ")
		prefix  string
		command string
		params  []string
		text    string
	)

	if strings.HasPrefix(message[0], ":") {
		prefix = extractPrefix(&message)
	}

	command = extractCommand(&message)

	for i, token := range message {
		if strings.HasPrefix(token, ":") {
			text = strings.Join(message[i:], " ")[1:]
			break
		}

		params = append(params, token)
	}

	return RawMessage{prefix, command, params, text}
}

func extractPrefix(message *[]string) string {
	prefix := (*message)[0]
	*message = (*message)[1:]

	return prefix
}

func extractCommand(message *[]string) string {
	command := (*message)[0]
	*message = (*message)[1:]

	return command
}

File: src/client/message.ts
export type RawMessage = {
  source: string
  prefix: string | undefined
  command: string
  params: string[]
  text: string | undefined
}

export function parseRawMessage(source: string): RawMessage {
  let message: Partial<RawMessage> = {}

  const tokens = source.trim().split(' ')
  const prefix = tokens[0]
  if (prefix && prefix.startsWith(':')) {
    message.prefix = prefix
    tokens.shift()
  }

  const command = tokens[0]
  if (command) {
    message.command = command
    tokens.shift()
  }

  message.params = []
  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i]
    if (token) {
      if (token.startsWith(':')) {
        message.text = tokens.slice(i).join(' ').substring(1)
        break
      }

      message.params.push(token)
    }
  }

  return buildRawMessage(source, message)
}

function buildRawMessage(source: string, message: Partial<RawMessage>): RawMessage {
  const { prefix, command, params, text } = message

  if (command === undefined) {
    throw new Error('failed to build message: command is undefined')
  }

  if (params === undefined) {
    throw new Error('failed to build message: params array is undefined')
  }

  return { source, prefix, command, params, text }
}
